File: datastructure/union-find/union_find.py
class UnionFind:
    def __init__(self, N=10):
        self.id = [i for i in range(N + 1)]
        self.sz = [1] * (N + 1)

    def root(self, i):
        while i != self.id[i]:
            self.id[i] = self.id[self.id[i]]
            i = self.id[i]
        return i

    def union(self, p, q):
        i = self.root(p)
        j = self.root(q)
        if self.sz[i] <= self.sz[j]:
            self.id[i] = j
            self.sz[j] += self.sz[i]
        else:
            self.id[j] = i
            self.sz[i] += self.sz[j]

    def connected(self, p, q):
        return self.root(p) == self.root(q)

File: datastructure/union-find/UnionFind.java

public class UnionFind {
    public int[] ids;
    public int[] sz;
    public int n;

    public UnionFind(int n) {
        this.n = n;
        ids = new int[n+1];
        sz = new int[n+1];
        for (int i=0;i<=n;i++) {
            ids[i] = i;
            sz[i] = 1;
        }
    }

    public void union(int p, int q) {
        p = root(p);
        q = root(q);
        if (sz[p] <= sz[q]){
            ids[p] = q;
            sz[q] += sz[p];
        }
        else{
            ids[q] = p;
            sz[p] += sz[q];
        }
    }

    public boolean isConnected(int p, int q) {
        return root(p) == root(q);
    }

    public int root(int x) {
        while (x != ids[x]) {
            ids[x] = ids[ids[x]];
            x = ids[x];
        }
        return x;
    }
}
